confignode/src/main/java/org/apache/iotdb/confignode/partition/PartitionTable.java

```diff
@@ -18,42 +18,169 @@
+
+    this.schemaLock = new ReentrantReadWriteLock();
+    this.schemaPartition = new SchemaPartitionInfo();
 
-    this.dataPartitionTable = new HashMap<>();
-    this.dataRegionDataNodesMap = new HashMap<>();
+    this.dataLock = new ReentrantReadWriteLock();
+    this.dataPartition = new DataPartitionInfo();
+  }
+
+  public TSStatus registerDataNode(RegisterDataNodePlan plan) {
+    TSStatus result;
+    DataNodeInfo info = plan.getInfo();
+    dataNodeLock.writeLock().lock();
+
+    if (dataNodesMap.containsValue(info)) {
+      result = new TSStatus(TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
+      result.setMessage(
+          String.format(
+              "DataNode %s is already registered.", plan.getInfo().getEndPoint().toString()));
+    } else {
+      info.setDataNodeID(nextDataNode);
+      dataNodesMap.put(info.getDataNodeID(), info);
+      result = new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
+      result.setMessage(String.valueOf(nextDataNode));
+      nextDataNode += 1;
+    }
 
-    this.dataPartitionRuleTable = new HashMap<>();
+    dataNodeLock.writeLock().unlock();
+    return result;
   }
 
-  // TODO: Interfaces for metadata operations
+  public Map<Integer, DataNodeInfo> getDataNodeInfo(QueryDataNodeInfoPlan plan) {
+    Map<Integer, DataNodeInfo> result = new HashMap<>();
+    dataNodeLock.readLock().lock();
 
-  // TODO: Interfaces for data operations
+    if (plan.getDataNodeID() == -1) {
+      result.putAll(dataNodesMap);
+    } else {
+      if (dataNodesMap.containsKey(plan.getDataNodeID())) {
+        result.put(plan.getDataNodeID(), dataNodesMap.get(plan.getDataNodeID()));
+      } else {
+        result = null;
+      }
+    }
+
+    dataNodeLock.readLock().unlock();
+    return result;
+  }
 
-  // TODO: Interfaces for data partition rules
+  public TSStatus setStorageGroup(SetStorageGroupPlan plan) {
+    TSStatus result;
+    storageGroupLock.writeLock().lock();
+
+    if (dataNodesMap.size() < regionReplicaCount) {
+      result = new TSStatus(TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
+      result.setMessage("DataNode is not enough, please register more.");
+    } else {
+      if (storageGroupsMap.containsKey(plan.getSchema().getName())) {
+        result = new TSStatus(TSStatusCode.INTERNAL_SERVER_ERROR.getStatusCode());
+        result.setMessage(
+            String.format("StorageGroup %s is already set.", plan.getSchema().getName()));
+      } else {
+        StorageGroupSchema schema = new StorageGroupSchema(plan.getSchema().getName());
+        regionAllocation(schema);
+        storageGroupsMap.put(schema.getName(), schema);
+        result = new TSStatus(TSStatusCode.SUCCESS_STATUS.getStatusCode());
+      }
+    }
+
+    storageGroupLock.writeLock().unlock();
+    return result;
+  }
+
+  private void regionAllocation(StorageGroupSchema schema) {
+    // TODO: 2PL may cause deadlock, remember to optimize
+    dataNodeLock.writeLock().lock();
```

Comment:
Why use so many locks... The PartitionTable can only be written by a single thread and read by multiple threads. It seems that a read/write lock is sufficient.

Reply:
Fixed.